src/container.jl

```diff
@@ -1,81 +1,79 @@
-mutable struct Trace{Tspl<:AbstractSampler, Tvi<:AbstractVarInfo, Tmodel<:Model}
-    model::Tmodel
-    spl::Tspl
-    vi::Tvi
-    ctask::CTask
-
-    function Trace{SampleFromPrior}(model::Model, spl::AbstractSampler, vi::AbstractVarInfo)
-        return new{SampleFromPrior,typeof(vi),typeof(model)}(model, SampleFromPrior(), vi)
-    end
-    function Trace{S}(model::Model, spl::S, vi::AbstractVarInfo) where S<:Sampler
-        return new{S,typeof(vi),typeof(model)}(model, spl, vi)
-    end
+struct Trace{F}
+    f::F
+    ctask::Libtask.CTask
 end
 
-function Base.copy(trace::Trace)
-    vi = deepcopy(trace.vi)
-    res = Trace{typeof(trace.spl)}(trace.model, trace.spl, vi)
-    res.ctask = copy(trace.ctask)
-    return res
-end
+const Particle = Trace
 
-# NOTE: this function is called by `forkr`
-function Trace(f, m::Model, spl::AbstractSampler, vi::AbstractVarInfo)
-    res = Trace{typeof(spl)}(m, spl, deepcopy(vi))
-    ctask = CTask() do
-        res = f()
-        produce(nothing)
-        return res
-    end
-    task = ctask.task
-    if task.storage === nothing
-        task.storage = IdDict()
+function Trace(f)
+    ctask = let f=f
+        Libtask.CTask() do
+            res = f()
+            Libtask.produce(nothing)
+            return res
+        end
     end
-    task.storage[:turing_trace] = res # create a backward reference in task_local_storage
-    res.ctask = ctask
-    return res
-end
 
-function Trace(m::Model, spl::AbstractSampler, vi::AbstractVarInfo)
-    res = Trace{typeof(spl)}(m, spl, deepcopy(vi))
-    reset_num_produce!(res.vi)
-    ctask = CTask() do
-        res = m(vi, spl)
-        produce(nothing)
-        return res
-    end
-    task = ctask.task
-    if task.storage === nothing
-        task.storage = IdDict()
-    end
-    task.storage[:turing_trace] = res # create a backward reference in task_local_storage
-    res.ctask = ctask
-    return res
+    # add backward reference
+    newtrace = Trace(f, ctask)
+    addreference!(ctask.task, newtrace)
+
+    return newtrace
 end
 
-# step to the next observe statement, return log likelihood
-Libtask.consume(t::Trace) = (increment_num_produce!(t.vi); consume(t.ctask))
+Base.copy(trace::Trace) = Trace(trace.f, copy(trace.ctask))
+
+# step to the next observe statement and
+# return the log probability of the transition (or nothing if done)
+advance!(t::Trace) = Libtask.consume(t.ctask)
```

Comment:
Yes, I agree. But I noticed that this would require additional changes since we have to propagate the state of the iterators in some way, if we do not assume that they are mutating. Hence I did not switch to the interface in this PR.